### PR TITLE
General function application, not just restricted to variables

### PR DESCRIPTION
**Pull request by Ekdohibs**
Also, now an error is produced for under-application of labelled functions.

### CODE REVIEW

**Review by gasche — APPROVED** [review]
The change is correct but, from a language design point of view, it's also slightly unpleasant: we only track shapes for variables (due to our untyped-semantics approach), so calling an arbitrary expression is a new source of incompleteness; I liked the old design where the restriction was explicit in the syntax. Oh well... I think it's fine this way.

**Comment by Ekdohibs** [comment]
Concerning calling an arbitrary expression being a source of incompleteness, I wanted to add a warning/error when a function with a non-trivial shape is used as a variable (and not applied at all). However, it's not obvious how to do it with the current scheme of compilation, as the warning would trigger when compiling the argument.

**Comment by Ekdohibs** [comment]
I pushed a commit to add an error when a function with a non-trivial shape escapes. However, we still have some incorrect behaviors caused by `align-args` not being always correct: if f is `let f ~x y = ...`, it will treat `f 2` as an application of a function when omitting labels, and behave as if `f ~x:2` had been written. I am not sure how to fix that, however.

File: miniml/compiler/compile.scm [hunk 1]
```diff
@@ -1779,18 +1796,25 @@
     (append h (list
       (list (list 'PWild) #nil (list 'EReraise (lid->evar exn-arg)))))))
 
+(define (lower-var env ld check-shape)
+  (let* ((vdef (env-get-var env ld))
+         (vloc (var-get-location vdef))
+         (vshape (var-get-funshape vdef)))
+    (if check-shape (for-each (lambda (lab) (assert (equal? (car lab) 'Nolabel))) vshape))
+    (match (var-get-location (env-get-var env ld))
```
Comment (gasche) [inline miniml/compiler/compile.scm:1804]: this should be `vloc`?

File: miniml/compiler/compile.scm [hunk 2]
```diff
@@ -1245,7 +1260,9 @@
           ))
   (define (align shape args)
     (cond ((null? shape) (if (null? args) #nil (align (map (lambda (arg) (list 'Nolabel)) args) args)))
-          ((null? args) #nil)
+          ((null? args) (begin
+                          (for-each (lambda (lab) (assert (equal? (car lab) 'Nolabel))) shape)
```
Comment (gasche) [inline miniml/compiler/compile.scm:1264]: this `for-each` comes up also in `lower-var`; maybe factorize it as `(check-no-labels shape)`?

**Review by gasche — APPROVED** [review]
Note: `align-args` is a part of the compiler that is barren land to me, I think trying to read the code would make my eyes bleed. On one hand, forcing myself to refactor it would be a sure way to understand it better, but on the other I don't like it when my eyes bleed so I decided to not read the code instead. You're basically on your own on this part -- which I think is fine.

**Comment by gasche** [comment]
@Ekdohibs I had made minor comments for simplifications above. Should I send a PR to propose them?

**Comment by Ekdohibs** [comment]
@gasche The changes are here: https://github.com/Ekdohibs/camlboot/commit/8bfbae567367c0a9768960f6db0a9faffd9569a4 .
If you had other suggestions, I'd be happy if you sent a PR!

**Comment by gasche** [comment]
Ah, I missed this commit, sorry. (Usually github warns when a comment is on an outdated code section and I relied on this.) Thanks!

**Pull request merged**